apis/privilegedaccessmanager/v1beta1/privilegedaccessmanagerentitlement_types.go

```diff
@@ -0,0 +1,315 @@
+}
+
+// AccessControlEntry is used to control who can do some operation.
+type AccessControlEntry struct {
+	// Optional. Users who are allowed for the operation. Each entry should be a
+	// valid v1 IAM principal identifier. The format for these is documented at:
+	// https://cloud.google.com/iam/docs/principal-identifiers#v1
+	// +optional
+	Principals *[]string `json:"principals,omitempty"`
+}
+
+// Different types of approval workflows that can be used to gate privileged
+// access granting.
+type ApprovalWorkflow struct {
+	// An approval workflow where users designated as approvers review and act
+	// on the grants.
+	// +required
+	ManualApprovals ManualApprovals `json:"manualApprovals,omitempty"`
+}
+
+// A manual approval workflow where users who are designated as approvers
+// need to call the 'ApproveGrant'/'DenyGrant' APIs for a grant. The workflow
+// can consist of multiple serial steps where each step defines who can act as
+// approver in that step and how many of those users should approve before the
+// workflow moves to the next step.
+//
+// This can be used to create approval workflows such as:
+//
+// * Require an approval from any user in a group G.
+// * Require an approval from any k number of users from a Group G.
+// * Require an approval from any user in a group G and then from a user U.
+//
+// A single user might be part of the 'approvers' ACL for multiple steps in this
+// workflow, but they can only approve once and that approval is only considered
+// to satisfy the approval step at which it was granted.
+type ManualApprovals struct {
+	// Optional. Do the approvers need to provide a justification for their
+	// actions?
+	// +optional
+	RequireApproverJustification *bool `json:"requireApproverJustification,omitempty"`
+
+	// Optional. List of approval steps in this workflow. These steps are followed
+	// in the specified order sequentially. Only 1 step is supported.
+	// +optional
+	Steps *[]Step `json:"step,omitempty"`
+}
+
+// Step represents a logical step in a manual approval workflow.
+type Step struct {
+	// Optional. The potential set of approvers in this step. This list must
+	// contain at most one entry.
+	// +optional
+	Approvers *[]AccessControlEntry `json:"approvers,omitempty"`
+
+	// Required. How many users from the above list need to approve. If there
+	// aren't enough distinct users in the list, then the workflow indefinitely
+	// blocks. Should always be greater than 0. 1 is the only supported value.
+	// +required
+	ApprovalsNeeded int32 `json:"approvalsNeeded,omitempty"`
+
+	// Optional. Additional email addresses to be notified when a grant is
+	// pending approval.
+	// +optional
+	ApproverEmailRecipients *[]string `json:"approverEmailRecipients,omitempty"`
+}
+
+// Privileged access that this service can be used to gate.
+type PrivilegedAccess struct {
+	// Access to a Google Cloud resource through IAM.
+	// +required
+	GcpIAMAccess GcpIamAccess `json:"gcpIAMAccess,omitempty"`
```

Comment:
Shall this be a ref to IAMPolicy?

Reply:
I don't think so. This resource is technically a IAMPolicyMember resource with additional restrictions. Referencing an IAMPolicy here means that the user need the IAM policy and a PAM entitlement containing the same policy with some restrictions. But these two resources have mutually exclusive functionatlities: If the policy can be configured, then the entitlement is not necessary; and if the entitlement is needed, then they shouldn't configure the policy.